### PR TITLE
Editor: Don't break shortcodes when they're selected in the editor

### PR DESCRIPTION
**Pull request by bisko**
This PR backports some of the WordPress.org core functionality that keeps the editor selection when switching between Visual and HTML modes. 

Originally reported in https://core.trac.wordpress.org/ticket/42908

The code in Calypso was an earlier version of the code that shipped in WordPress.org and didn't take in account how shortcodes work. This resulted in the selection marker being inserted in the middle of a shortcode, which broke shortcodes that had live rendering enabled for them (e.g. Gallery).

The way it was handled in Core was to move the cursor/s to the edge of the shortcode, if it was inside a shortcode. This way the whole shortcode is selected, to avoid inserting the marker inside the shortcode and therefore breaking it when switching to Visual mode.

The backport brings only the functionality to move the cursor to select the whole shortcode. See Known Issues below for a slight regression. 

To test:
1. Create post with a Gallery or another shortcode
2. Switch to HTML mode
3. Select some part of the Gallery shortcode (e.g. click in the middle of it or select some of the image ids)
4. Switch to Visual mode
5. Verify the Gallery is rendered properly, without breaking. 

Known issues:
Right now when a selection is made in HTML mode, switched to Visual and there are embeds in the content, the selection will be completely lost. That's an unfortunate side effect from the code that loads embeds and can be addressed in a separate PR.

cc @lizkarkoski @azaozz

### CODE REVIEW

File: client/post-editor/post-editor.jsx [hunk 1]
```diff
@@ -69,6 +69,7 @@ import { isSitePreviewable } from 'state/sites/selectors';
 import { removep } from 'lib/formatting';
 import QuickSaveButtons from 'post-editor/editor-ground-control/quick-save-buttons';
 import EditorRevisionsDialog from 'post-editor/editor-revisions/dialog';
+import { regexp as ShortCodeRegexp } from 'lib/shortcode';
```
Comment (jblz) [inline client/post-editor/post-editor.jsx]: `ShortCodeRegexp` be lowercase, right?

Comment (bisko) [reply client/post-editor/post-editor.jsx]: Good point! Updated.

**Comment by azaozz** [comment]
> Right now when a selection is made in HTML mode, switched to Visual and there are embeds in the content, the selection will be completely lost.

Yeah, if there is a `wpview` somewhere in the editor the selection and scrolling to its position are lost when switching to Visual, regardless of where the cursor is. This is not a regression, also happens without this PR.

The rest seems to be working well, the shortcodes are intact.

File: client/post-editor/post-editor.jsx [hunk 2]
```diff
@@ -69,6 +69,7 @@ import { isSitePreviewable } from 'state/sites/selectors';
 import { removep } from 'lib/formatting';
 import QuickSaveButtons from 'post-editor/editor-ground-control/quick-save-buttons';
 import EditorRevisionsDialog from 'post-editor/editor-revisions/dialog';
+import { regexp as shortCodeRegexp } from 'lib/shortcode';
```
Comment (Copons) [inline client/post-editor/post-editor.jsx]: Please notice (here and in the other occurrences in this change) that "shortcode" does not have the capital C.
See: https://codex.wordpress.org/Shortcode_API

**Comment by alisterscott** [comment]
This branch needs a rebase. Is this PR still required?

**Comment by alisterscott** [comment]
Closing due to inactivity - please reopen if necessary

**Pull request closed without merging**